{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3799e922-f05d-44b1-8162-e3d0d5aa0f37",
   "metadata": {},
   "source": [
    "# Motivation\n",
    "\n",
    "The NCBI Cell pilot seeks to demonstrate the feasibility of developing a\n",
    "cell phenotype knowledgebase resource that captures knowledge about cell\n",
    "phenotypes, exposes this knowledge for search and analysis by the\n",
    "research community, and integrates the cell phenotype knowledge with\n",
    "other sources of knowledge about diseases and drugs to enable discovery\n",
    "of novel diagnostic biomarkers and therapeutic targets.\n",
    "\n",
    "Two streams of knowledge are envisioned:\n",
    "\n",
    "-   Single cell genomics data captured from existing data repositories\n",
    "    (e.g., CZI CELLxGENE, NeMO) will be analyzed to produce information\n",
    "    about cell type-specific marker genes and differential expression\n",
    "    patterns\n",
    "\n",
    "-   Open access peer-reviewed publications reporting results from single\n",
    "    cell genomics experiments from PubMed Central (PMC) will be\n",
    "    processed to extract information about cell type-specific marker\n",
    "    genes and differential expression patterns\n",
    "\n",
    "In both cases, the information will be linked with experiment metadata\n",
    "about species and specimen sample sources, disease states, and\n",
    "perturbation responses.\n",
    "\n",
    "Furthermore, in order to maximize the interoperability of this derived\n",
    "knowledge with other sources of knowledge about genes, pathways,\n",
    "diseases, and drugs, the information will be in the form of standardized\n",
    "semantically-structured assertions of subject-predicate-object triple\n",
    "statements that will be compatible with storage using semantic web\n",
    "technologies (e.g., RDF, OWL) and graph database (e.g., ArangoDB)\n",
    "platforms.\n",
    "\n",
    "# Goals\n",
    "\n",
    "As a beginning step of the pilot, this document describes use of Python\n",
    "to demonstrate both envisioned knowledge streams, and simplified use of\n",
    "a graph database for visualization. In particular, goals of this\n",
    "document include:\n",
    "\n",
    "-   Identification of CELLxGENE datasets and their citations within\n",
    "    PubMed (See\n",
    "    [Chapter-01-CELLxGENE.ipynb](Chapter-01-CELLxGENE.ipynb))\n",
    "\n",
    "-   Use of package NSForest for the discovery of minimum marker gene\n",
    "    combinations for cell type identification from single-cell RNA\n",
    "    sequencing data (See\n",
    "    [Chapter-02-NSforest.ipynb](Chapter-02-NSForest.ipynb))\n",
    "\n",
    "-   Use of package OntoGPT for extracting structured information from\n",
    "    text with large language models (LLMs), instruction prompts, and\n",
    "    ontology-based grounding (See\n",
    "    [Chapter-03-ontoGPT.ipynb](Chapter-03-OntoGPT.ipynb))\n",
    "\n",
    "-   Creation, population, and display of the processing results using\n",
    "    graph database ArangoDB (See\n",
    "    [Chapter-04-ArangoDB.ipynb](Chapter-04-ArangoDB.ipynb))\n",
    "\n",
    "# Development environment\n",
    "\n",
    "A development environment requires a few environment variables, as well\n",
    "as a virtual environment containing Python dependencies.\n",
    "\n",
    "## Environment variables\n",
    "\n",
    "Accessing PubMed without limit requires creating an [API\n",
    "key](https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/)\n",
    "associated with the email address of an NCBI account. Running OntoGPT\n",
    "requires creating and funding an [OpenAI\n",
    "platform](https://openai.com/api/) account, configuring a [project API\n",
    "key](https://platform.openai.com/api-keys) , and creating a\n",
    "[BioPortal](https://bioportal.bioontology.org/login) account to be\n",
    "assigned an API key.\n",
    "\n",
    "Since these keys should not be made public, this demonstration assigns\n",
    "them to environment variables in a `.zshenv` file as follows:\n",
    "\n",
    "``` zsh\n",
    "export BIOPORTAL_API_KEY=<YOUR_BIOPORTAL_API_KEY>\n",
    "export OPENAI_API_KEY=<YOUR_OPENAI_API_KEY>\n",
    "export NCBI_EMAIL=<YOUR_NCBI_EMAIL>\n",
    "export NCBI_API_KEY=<YOUR_NCBI_API_KEY>\n",
    "```\n",
    "\n",
    "This file must be created and sourced prior to running this\n",
    "demonstration. Do not commit this file to the repository.\n",
    "\n",
    "## Python dependencies\n",
    "\n",
    "This demonstration uses [Poetry](https://python-poetry.org/) to manage\n",
    "dependencies. Poetry was installed in its own virtual environment\n",
    "following the recommended (manual) approach as follows\n",
    "\n",
    "``` zsh\n",
    "python -m venv .poetry\n",
    "source .poetry/bin/activate\n",
    "python -m pip install -i .poetry.txt\n",
    "deactivate\n",
    "```\n",
    "\n",
    "Then the dependencies specified in the `pyproject.toml` file in the root\n",
    "directory of this repository were installed as follows:\n",
    "\n",
    "``` zsh\n",
    "python -m venv .venv\n",
    "source .venv/bin/activate\n",
    ".poetry/bin/poetry install\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
